src/OrleansRuntime/Streams/QueueBalancer/LeaseBasedQueueBalancer.cs

```diff
@@ -0,0 +1,345 @@
+﻿using Orleans.LeaseProviders;
+using Orleans.Runtime;
+using Orleans.Runtime.Configuration;
+using Orleans.Streams;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Orleans.Streams
+{
+    /// <summary>
+    /// IResourceSelector selects a centain amount of resource from a resource list
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    internal interface IResourceSelector<T>
+    {
+        /// <summary>
+        /// Try to select certain count of resources from resource list, which doesn't overlap with existing selection
+        /// </summary>
+        /// <param name="newSelectionCount"></param>
+        /// <param name="existingSelection"></param>
+        /// <returns></returns>
+        IEnumerable<T> NextSelection(int newSelectionCount, IEnumerable<T> existingSelection);
+    }
+
+    /// <summary>
+    /// Selector using round robin algorithm
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    internal class RoundRobinSelector<T> : IResourceSelector<T>
+    {
+        private List<T> resources;
+        private int lastSelection;
+        public RoundRobinSelector(List<T> resources)
+        {
+            this.resources = resources;
+            this.lastSelection = -1;
+        }
+
+        /// <summary>
+        /// Try to select certain count of resources from resource list, which doesn't overlap with existing resources
+        /// </summary>
+        /// <param name="newSelectionCount"></param>
+        /// <param name="existingSelection"></param>
+        /// <returns></returns>
+        public IEnumerable<T> NextSelection(int newSelectionCount, IEnumerable<T> existingSelection)
+        {
+            var selection = new List<T>();
+            while (selection.Count < newSelectionCount)
+            {
+                this.lastSelection = (++this.lastSelection) % (this.resources.Count - 1);
+                if(!existingSelection.Contains(this.resources[this.lastSelection]))
+                    selection.Add(this.resources[this.lastSelection]);
+            }
+            return selection;
+        }
+    }
+
+    public class AzureDeploymentLeaseBasedBalancer : LeaseBasedQueueBalancer
+    {
+        public AzureDeploymentLeaseBasedBalancer(ILeaseProvider leaseProvider, ISiloStatusOracle siloStatusOracle,
+            IServiceProvider serviceProvider)
+            : base(leaseProvider, siloStatusOracle, DeploymentBasedQueueBalancerUtils.CreateDeploymentConfigForAzure(serviceProvider),
+                  serviceProvider.GetRequiredService<Factory<string, Logger>>())
```

Comment:
`Factory<string, Logger>` can be injected directly into the ctor, instead of injecting `IServiceProvider` (which is usually considered an anti-pattern, even though we do it in some places out of necessity) and pulling it from there.